### PR TITLE
Introduce E2E tests for the faucet bot and server

### PR DESCRIPTION
**Pull request by rzadp**
First, we spin up:
- A local Polkadot node in a dev mode
- A local matrix server
  - Default configuration, except for increased rate limits
  - We create necessary users and gather them in a room
- The faucet backend
  - It uses `//Alice` seed so it is funded
- The faucet bot

Next, we test the behaviour by sending `!balance` and `!drip` messages.

### CODE REVIEW

File: .github/workflows/E2E.yml [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+      - main
+jobs:
+  e2e:
+    timeout-minutes: 15
+    runs-on: ubuntu-22.04
+
+    steps:
+      - uses: actions/checkout@v3.3.0
+      - name: Install prerequisites
+        run: sudo apt-get update && sudo apt-get install -y jq
```
Comment (Bullrich) [inline .github/workflows/E2E.yml:15]: `jq` is [already bundled in ubuntu-22.04](https://github.com/actions/runner-images/blob/main/images/linux/Ubuntu2204-Readme.md)

**Review by Bullrich — CHANGES_REQUESTED** [review]
Looks to me.

But seriously, good pr. **Lots** of setups I see.

Should we also add tests about the bot interacting with invalid addresses?

File: .github/workflows/E2E.yml (hunk 1 above)
Comment (rzadp) [reply .github/workflows/E2E.yml]: Thanks, good to know - removed the install.

**Comment by rzadp** [comment]
> Should we also add tests about the bot interacting with invalid addresses?

Yes, I would be adding some more test cases in a separate PR.

**Review by mordamax — APPROVED** [review]

**Review by Bullrich — APPROVED** [review]

File: src/utils.ts [hunk 2]
```diff
@@ -1,3 +1,20 @@
...
+  const stopTime = Date.now() + timeout;
+  while (Date.now() <= stopTime) {
+    const result = await predicate();
+    if (result) {
+      return result;
+    }
+    await sleep(step);
+  }
+  throw new Error(`waitUntil timed out after ${timeout} ms`);
+};
```
Comment (mutantcornholio) [inline src/utils.ts:20]: There are already `delay` and and `until` functions in `opstooling-js`

**Review by mutantcornholio — CHANGES_REQUESTED** [review]
Small nit about reusing common code, still going through test setup configuration

File: src/utils.ts [hunk 3]
```diff
@@ -1,3 +1,20 @@
 export function isAccountPrivileged(sender: string): boolean {
   return sender.endsWith(":parity.io") || sender.endsWith(":web3.foundation");
 }
+
+export const sleep = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+type Predicate = () => any | (() => Promise<any>);
+
+export const waitUntil = async (predicate: Predicate, step = 100, timeout = 10000) => {
```
Comment (mutantcornholio) [inline src/utils.ts:10]: Also, the timeout here's the same as the governing timeout for the tests, meaning that this timeout condition will never fire, and instead we'll be getting timeouts from jest, without stacktrace

**Comment by PierreBesson** [comment]
IMO setting up docker containers using docker exec and wait-for in bash scripts can be flaky. I would recommend using a proper docker test framework such as https://github.com/testcontainers/testcontainers-node

**Comment by mutantcornholio** [comment]
> IMO setting up docker containers using docker exec and wait-for in bash scripts can be flaky. I would recommend using a proper docker test framework such as https://github.com/testcontainers/testcontainers-node

Good point actually. This advice could save me quite some time a month ago :)

**Comment by rzadp** [comment]
> IMO setting up docker containers using docker exec and wait-for in bash scripts can be flaky. I would recommend using a proper docker test framework such as https://github.com/testcontainers/testcontainers-node

I had no idea this existed, I will check it out.

**Comment by Bullrich** [comment]
> IMO setting up docker containers using docker exec and wait-for in bash scripts can be flaky. I would recommend using a proper docker test framework such as https://github.com/testcontainers/testcontainers-node

That looks super useful! Will keep it in mind

**Review by mutantcornholio — APPROVED** [review]

**Review by sergejparity — APPROVED** [review]

File: src/utils.ts (hunk 2 above)
Comment (rzadp) [reply src/utils.ts]: Thanks for pointing that out - applied.

File: src/utils.ts (hunk 3 above)
Comment (rzadp) [reply src/utils.ts]: Good catch, timeouts adjusted.

File: e2e/bootstrap.sh [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+# Clean up potential data from previous runs,
+# in order to start with a clean state.
+docker-compose down -v
+rm -rf ./matrix_data/homeserver.db* ../sqlite.db
+
+
+# Start Polkadot and Matrix and wait until they are up.
+docker-compose up -d
+source wait_until.sh 'curl -s "localhost:8008"'
+source wait_until.sh 'curl -s "localhost:9933"'
```
Comment (parutger) [inline e2e/bootstrap.sh:14]: This should technically be 127.0.0.1 instead of localhost

Comment (rzadp) [reply e2e/bootstrap.sh]: Ok, updated.

**Pull request merged**